src/main/java/io/split/android/client/cache/MySegmentsCache.java

```diff
@@ -1,72 +1,103 @@
+import android.arch.lifecycle.LifecycleObserver;
+import android.arch.lifecycle.OnLifecycleEvent;
+import android.arch.lifecycle.ProcessLifecycleOwner;
 
-import io.split.android.client.api.Key;
 import io.split.android.client.dtos.MySegment;
 import io.split.android.client.storage.IStorage;
 import io.split.android.client.utils.Json;
 import io.split.android.client.utils.Logger;
+import io.split.android.client.utils.Utils;
 
 /**
  * Created by guillermo on 12/28/17.
  */
 
-public class MySegmentsCache implements IMySegmentsCache {
+public class MySegmentsCache implements IMySegmentsCache, LifecycleObserver {
 
-    private static final String MY_SEGMENTS_FILE_PREFIX = "SPLITIO.mysegments";
+    private static final String MY_SEGMENTS_FILE_NAME = "SPLITIO.mysegments";
 
-    private final IStorage _storage;
+    private final IStorage mFileStorage;
+    private Map<String, List<MySegment>> mSegments = null;
 
     public MySegmentsCache(IStorage storage) {
-        _storage = storage;
+        ProcessLifecycleOwner.get().getLifecycle().addObserver(this);
+        mFileStorage = storage;
+        mSegments = Collections.synchronizedMap(new HashMap<String, List<MySegment>>());
+        loadSegmentsFromDisk();
     }
 
-    private String getMySegmentsId() {
-        return MY_SEGMENTS_FILE_PREFIX;
+    private String getMySegmentsFileName() {
+        return MY_SEGMENTS_FILE_NAME;
     }
 
-    private String getMySegmentsKeyId() {
-        return MY_SEGMENTS_FILE_PREFIX + ".currentKey";
+    @Override
+    public void setMySegments(String key, List<MySegment> mySegments) {
+        mSegments.put(key, mySegments);
     }
 
     @Override
-    public boolean saveMySegments(String key, List<MySegment> mySegments) {
-        try {
-            _storage.write(getMySegmentsKeyId(), key);
-            _storage.write(getMySegmentsId(), Json.toJson(mySegments));
-            return true;
-        } catch (IOException e) {
-            Logger.e(e, "Could not save my segments");
-            return false;
-        }
+    public List<MySegment> getMySegments(String key) {
+        return mSegments.get(key);
     }
 
     @Override
-    public List<MySegment> getMySegments(String key) {
+    public void deleteMySegments(String key) {
+        mSegments.remove(key);
+    }
+
+    private void loadSegmentsFromDisk(){
+
         try {
-            String savedKey = _storage.read(getMySegmentsKeyId());
-            if (savedKey != null && savedKey.equals(key)) {
-                String storedMySegments = _storage.read(getMySegmentsId());
-                Type listType = new TypeToken<List<MySegment>>() {
-                }.getType();
-
-                return Json.fromJson(storedMySegments, listType);
-            } else {
-                _storage.delete(getMySegmentsId());
+            String storedMySegments = mFileStorage.read(getMySegmentsFileName());
+            if(storedMySegments == null || storedMySegments.trim().equals("")) return;
```

Comment:
nit, I k now it is annoying

if (storedMySegments == null || storedMySegments.trim().equals("")) {
return
}